image/icns.ksy

```diff
@@ -0,0 +1,889 @@
+  depending on the icon's displayed size and the capabilities of the display hardware.
+
+  An icon family in an ICNS file can also contain other nested icon families.
+  This feature is sometimes used to represent different states or visual variants of the same icon,
+  such as an opened, selected or dark mode version.
+
+  Each icon in an icon family is identified by a four-byte type code,
+  which also indicates the image data format and resolution.
+  Supported formats include PNG and JPEG 2000 (for larger sizes and modern systems)
+  and multiple raw bitmap formats with varying resolution, color depth, and transparency support
+  (for smaller sizes and/or compatibility with older systems).
+
+  ICNS data can be stored either as a standalone file,
+  or as a resource with type code `'icns'` in a resource fork.
+  The latter was especially common on Classic Mac OS and for Carbon applications,
+  and is still used (as of macOS 10.14) by the Finder to store custom file and folder icons set by the user.
+doc-ref:
+  - '<OSServices/IconStorage.h>'
+  - '<OSServices/IconStorage.r>'
+  - '<HIServices/Icons.r>'
+  - 'https://sourceforge.net/p/icns/ libicns SourceForge project'
+  - 'https://github.com/python-pillow/Pillow/blob/master/src/PIL/IcnsImagePlugin.py Python Pillow ICNS plugin code'
+  - "https://www.macdisk.com/maciconen.php#icns Information about 'icns' resources in resource forks"
+seq:
+  - id: root_element
+    type: icon_family_element
+    valid:
+      expr: |
+        _.header.type.as_enum == icon_family_element::header::type::main_family
+    doc: |
+      The element at the root of the file.
+      This element must have type `"icns"`,
+      should be as long as the entire stream,
+      and contains the main icon family.
+types:
+  icon_family_element:
+    doc: |
+      A single element in an icon family.
+
+      It is normally safe to ignore elements with an unrecognized type code or unparseable data when reading -
+      they are most likely new icon types/resolutions, icon family variants, or metadata introduced in a newer system version.
+      However, when modifying and writing an icon family,
+      such elements should be stripped out,
+      to avoid leaving outdated information in the icon family that could be used by newer systems.
+    -webide-representation: '{header}'
+    seq:
+      - id: header
+        type: header
+        doc: The header storing the element's type code and length.
+      - id: data_with_io
+        type: bytes_with_io
+        size: header.len_data
+        doc: |
+          Use `data` instead,
+          unless you need access to this field's `_io`.
+    instances:
+      data:
+        value: data_with_io.data
+        doc: The raw data stored in the element.
+      data_parsed:
+        io: data_with_io._io
+        pos: 0
+        type:
+          switch-on: header.type.as_enum
+          cases:
+            'header::type::main_family': icon_family_data
```

Comment:
Given that they are text strings of fixed length and have a kind of structure within them .... would it worth not to use enums for them, but to decompose and parse them in KS?

Reply:
Yes, the type codes follow a certain pattern, but it's not consistent enough for parsing to make sense or work properly. The most predictable type codes are the ones for the Classic Mac OS-style bitmaps, which all follow the `ic□#`/`ic□4`/`ic□8`/`i□32`/`□8mk` naming scheme - except:

* the `m` ("mini") size doesn't have an RGB version (`im32`/`m8mk` don't exist)
* the `l` ("large") size the 1-bit+mask variant has type code `ICN#` (not `icl#`)
* the `t` ("thumbnail") size *only* has an RGB version (`ict#`/`ict4`/`ict8` don't exist)
* `it32` has a *slightly* different structure (four extra zero bytes at the beginning) than all the other `i□32` types

And it doesn't get much better with the PNG/JP2 types, which are called `ic0□` (regular scale) and `ic1□` (HiDPI scale), except:

* `ic04` (16x16) and `ic05` (32x32) are actually ARGB bitmaps - their PNG/JP2 versions are called `icp4` and `icp5`
* there's also `icp6`, which is 64x64 - a size that doesn't exist in *any* other type (it doesn't even have an ARGB or HiDPI version)
* `ic10` is the lowest-numbered HiDPI icon, but has the highest resolution (512x512) - because older OSes actually considered a non-HiDPI 1024x1024 icon

There's also `icsb`/`icsB`, which are ARGB in another odd size (18x18). And then you have the metadata and icon family types, which follow no naming pattern at all - one of them isn't even readable text (`0xfdd92fa8`)!

Reply:
Thanks. It would be nice to write content this summary into the spec itself.

Reply:
> It would be nice to write content this summary into the spec itself.

This is just a summary of what's already written in the spec - all the available type codes are listed in the enum, their format/structure is documented in the big switch and the `*_data` types, and the types that would fit the naming scheme but don't exist are pointed out in comments in between the real types.

> I also wonder, if they really don't exist or just undocumented. I mean, have anyone done any experiments on feeding the data with "nonexistent" ids into Apple software and observing the results?

They really don't exist as far as I can tell. I modified some icon files to include those icon types (`im32`, `m8mk`, `ict#`, `ict4`, `ict8`, `ic06`) on system versions that understand other similar icon types, and none of them show up in the Preview application. Actually the valid "mini" sizes don't show up under Preview either (even on Mac OS X 10.4) - to view those I had to use an older tool called icns Browser, and that clearly indicates that "mini" RGB and mask don't exist (it doesn't even show empty placeholders for those types, which it does for types that are valid but not in the file).

Reply:
Thanks.